### PR TITLE
Fixed branch name acquiring

### PR DESCRIPTION
**Pull request by AnrDaemon**
Read branch name from internal Git API if available, else read the output of git command.
This improves, for example, the support for Git workspaces without any extra work.

### CODE REVIEW

**Comment by AnrDaemon** [comment]
Looks like I've reimplemented #22 on a different level.

**Comment by koenlek** [comment]
Any updates on this? Would love to see this merged if it fixes #12

**Comment by StK88** [comment]
Hi,

About vscode git extension:

- Is this the usual way to call another extension? Is this the usual way to depend on another extension?
- How can we ensure compatibilty? Any way to list the dependency in a formal way?

[Git extension types file](https://github.com/wisetime-io/vscode-branch-in-window-title/pull/23/files#diff-8788f099bb5eddbd5fecbe781fad2bc200098924bb973d64447fcc9999c4d06d) is added manually, in case of git extension API changes it might become incompatible.

Does https://github.com/wisetime-io/vscode-branch-in-window-title/pull/22 solve your issues? [Commented](https://github.com/wisetime-io/vscode-branch-in-window-title/pull/22#issuecomment-2049724136) it as well.

**Comment by AnrDaemon** [comment]
Yes, it is an official way of calling Git extension. In its absence, there's a fallback of direct call to "git" executable.
Types file is added for easier development. I could have excluded it from commit, but that would only make it harder for those who would want to work with it in the future.
If anything changes in the future, we'd just plop a new definitions file in there and check for any discrepancy.

**Comment by AnrDaemon** [comment]
And yes, it solved my issue of multiple workspaces for the same repo.

**Comment by StK88** [comment]
Thanks for the clarifications.
There's no problem with having a git extension types file, the main concern is if the API/types of git extension will change, there's a low probability of a run time error from git extension without reaching the native git fallback. The current approach looks safe since at least `extension.exports.getAPI(1)` specifies the API version number, but in this case, we need to rely on backward-compatibility of future updates in https://github.com/microsoft/vscode/blob/main/extensions/git/src/api/git.d.ts.

Is there a significant benefit of involving git extension?
Theoretically, git extension and native git fix the same issues, but git extension might require manual sync/changes in future.

Minor note, branchDetector tests are failing:
Screenshot 2024-04-10 at 11 32 50
but I wouldn't like you to spend more time on it if we will go with https://github.com/wisetime-io/vscode-branch-in-window-title/pull/22 which uses only native git and solves same issues.

**Comment by AnrDaemon** [comment]
Accessing API is an order of magnitude faster compared to spawning git executable.
Also, in case of web-version and remote development, extension could be the only available way of accessing the repo.

**Comment by AnrDaemon** [comment]
I'm not 100% sure I fixed all tests. If you could give me a hand in this, your help would be greatly appreciated.
As it is, I'm not a JS/TS developer.

**Comment by koenlek** [comment]
@StK88 , FWIW: all my work is on remote systems and this change would finally make this extension usable for me (I've been wanting this for a long time!). 

Also, on the extension vs native git call discussion: arguably, going the extension route is perhaps more idiomatic than using native git, assuming the extension is binding git library APIs, whereas native git presumably runs it as a system call (shell command), which can be considered an anti-pattern (but without a doubt different people will have different opinions on that).

**Comment by StK88** [comment]
@AnrDaemon @koenlek, thanks for the feedback. Extended implementation with git extension makes sense. I had a look into failing tests and couldn't find a quick way to fix them. I'm unable to dedicate more time to fixing tests. https://github.com/wisetime-io/vscode-branch-in-window-title/pull/22 can be used as a tip/reference with fixed tests.

**Comment by StK88** [comment]
@AnrDaemon, can you temporarily give me permission to your fork, so I can push tests fix? Or I can fork your fork and open a separate PR.

**Comment by StK88** [comment]
ping @AnrDaemon

**Comment by AnrDaemon** [comment]
You can PR them to my repo. IDK about giving access.

**Comment by StK88** [comment]
@AnrDaemon opened https://github.com/AnrDaemon/vscode-branch-in-window-title/pull/1 can you please review and merge if all is okay?

**Comment by AnrDaemon** [comment]
I've had a look over it. We'll need an approval for new dependency first.

**Comment by vyshane** [comment]
The new dev dependency is fine.

**Comment by AnrDaemon** [comment]
I think we've resolved the testing issues. Please review.

File: package.json [hunk 1]
```diff
@@ -1,77 +1,79 @@
...
-		"mocha": "^9.2.2",
-		"text-encoding": "^0.7.0",
-		"typescript": "^4.6.3",
-		"vscode-test": "^1.6.1"
-	}
+    "name": "branch-in-window-title",
+    "displayName": "Branch in Window Title",
+    "description": "Visual Studio Code extension that shows the current Git branch in the window title. This can be used to automate time allocation to issue tracking systems via WiseTime.",
+    "publisher": "wisetime",
+    "version": "0.1.5",
```
Comment (StK88) [inline package.json:6]:
```suggestion
    "version": "0.2.0",
```

File: package.json [hunk 2]
```diff
@@ -1,77 +1,79 @@
...
+    "license": "Apache-2.0",
+    "repository": {
+        "type": "git",
+        "url": "https://github.com/wisetime-io/vscode-branch-in-window-title.git"
+    },
+    "engines": {
+        "vscode": "^1.70.0"
```
Comment (StK88) [inline package.json:13]: Can you revert `package.json` formatting tabs instead of spaces? It's hard to see actual changes.
Also, can you revert this change:
```suggestion
        "vscode": "^1.73.0"
```

**Review by StK88 — CHANGES_REQUESTED** [review]
A couple of minor things left and we're ready to merge.
To avoid confusion, I'm a responsible reviewer for this update.

File: package.json (hunk 2 above)
Comment (AnrDaemon) [reply package.json]: I've pushed `.editorconfig` to avoid future confusions about formatting of those files.

What to VSCode version base, it was changed around the time of #17 and I frankly don't understand the reasoning.
This unwarranted change excluded people using Windows 7 from benefitting by these updates.

**Comment by StK88** [comment]
@AnrDaemon,
>What to VSCode version base, it was changed around the time of https://github.com/wisetime-io/vscode-branch-in-window-title/pull/17 and I frankly don't understand the reasoning.
This unwarranted change excluded people using Windows 7 from benefitting by these updates.

Sorry, I don't think supporting Windows 7 is reasonable.
Microsoft support for Windows 7 ended in 2020.
By keeping the old vscode engine we miss all improvements and bug fixes from newer versions.
`1.73.0` is already quite outdated and we will update it in the future anyway. From history, we update it alongside other dependencies.
Are you on Windows 7? I'm not sure, maybe Windows version diff was a reason behind [test failures](https://github.com/AnrDaemon/vscode-branch-in-window-title/pull/1#issuecomment-2086403937) since our node versions are quite the same. I'm on Win 10.

**Comment by AnrDaemon** [comment]
Dropping support without a reason is unreasonable. Extension does not depend on any new API's, bumping dependency version "just because" is a very bad idea in general.

The extension does not package the "old engine", it only contains the code it contains. Which is easily proven by `npx vsce ls`.

**Comment by StK88** [comment]
Yes, it's simply a compatible vscode version.

There are a few reasons:

- VSCode [dropped support](https://code.visualstudio.com/docs/supporting/FAQ#_can-i-run-vs-code-on-older-windows-versions) for Windows 7/8/8.1
- we can't guarantee support for too old versions including a freeze of compatible VSCode version to `^1.70.0`
- if we receive a bug specific to too old versions, we won't spend time on it

Even if we leave `1.70.0` downgrade, we can't guarantee that it won't be changed in the future, therefore I asked for a revert, to avoid temporary false expectations and unrelated changes.

**Comment by AnrDaemon** [comment]
`^1.70.0` is not "freeze". It's "1.70.0 <= version < 2.0". So, until VSCode v2 it will be good.

**Review by StK88 — APPROVED** [review]
By freeze, I meant that now we must remember and restrict changes to the engine version due to Windows 7. Changes to implementation/tests will require regression or waiting for a fix from someone with Windows 7.

We've discussed it internally, we're fine to keep this change on engine with 1.70.0.
But still, we can't guarantee that it won't be changed in future.

Thanks for your contribution.

**Comment by AnrDaemon** [comment]
Thanks for your kind words and your understanding.
Is there anything left to attend?

**Comment by StK88** [comment]
All is good, thanks for the heads up.
We'll merge and release it as soon as possible.
